### PR TITLE
Screed Record Slicing

### PR DESCRIPTION
**Pull request by wrightmhw**
For https://github.com/ged-lab/khmer/issues/768

opening this pull request automatically redirected me to:
https://github.com/ctb/screed/compare/master...ged-lab:record_slicing?expand=1 (CTB's branch, not /ged-lab/screed)
Is there a reason for that? I switched to ged-lab

@ctb, I accidentally committed this change to master instead of opening this new branch, can you possibly delete that commit?

### CODE REVIEW

**Comment by bocajnotnef** [comment]
For whatever reason PRs always default to CTB's branch

**Comment by wrightmhw** [comment]
Thanks, @bocajnotnef :+1:

File: screed/screedRecord.py [hunk 1]
```diff
@@ -32,10 +32,18 @@ def __len__(self):
     def keys(self):
         return self.d.keys()
 
-    def __getitem__(self, sliced):
-        slicedkeys = self.d.keys()[sliced]
-        d = {i: self.d[i] for key in slicedkeys} # create a new dict from the sliked keys and return the dict
-        return d
+    def __getitem__(self, idx):
+        if isinstance(idx, slice):
```
Comment (wrightmhw) [inline screed/screedRecord.py]: @camillescott is this preferable to `if type(idx) is slice:`?

Comment (camillescott) [inline screed/screedRecord.py]: Yes, I think `isinstance` is better here -- it accounts for inheritance.

File: screed/tests/test_fasta.py [hunk 2]
```diff
@@ -145,3 +145,10 @@ class FakeRecord(object):
     w.consume(read_iter)
 
     assert fp.getvalue() == '>foo bar\nATCG\n'
+
+def test_slicing():
+    index = screed.open('screed/tests/test.fa')
```
Comment (wrightmhw) [inline screed/tests/test_fasta.py]: @mr-c Is this an appropriate way to open a file for a nosetest or should I use a 
`self._testfa = os.path.join(os.path.dirname(__file__), 'test.fa')`?

**Comment by bocajnotnef** [comment]
Will be relevant to https://github.com/ged-lab/screed/pull/30

**Comment by wrightmhw** [comment]
- [x] Is it mergeable?
- [x] Did it pass the tests?
- [x] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [x] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?

**Comment by wrightmhw** [comment]
@mr-c ready for review

File: screed/tests/test_fasta.py [hunk 3]
```diff
@@ -145,3 +145,28 @@ class FakeRecord(object):
...
+
+    def setup(self):
+        self._testfa = os.path.join(os.path.dirname(__file__),
+                                    'test.fa')
+        self._index = screed.open(self._testfa)
+        self._record = self._index.next()
+
+    def test_slicing(self):
+        test_slice = self._record[:10]['name']
+        assert test_slice == "ENSMICT00000012722"
```
Comment (mr-c) [inline screed/tests/test_fasta.py]: Can we get asserts on the content of the sequence data here?

File: screed/tests/test_fasta.py [hunk 4]
```diff
@@ -145,3 +145,28 @@ class FakeRecord(object):
...
+
+class Test_fastq_slicing(object):
+    def setup(self):
+        self._testfq = os.path.join(os.path.dirname(__file__),
+                                    'test.fastq')
+        self._index = screed.open(self._testfq)
+
+    def test_slicing(self):
+        self._record = self._index.next()
+        test_slice = self._record['quality'][:10]
```
Comment (mr-c) [inline screed/tests/test_fasta.py]: Like above, can we get asserts on the content of the sequence data here?

File: screed/tests/test_fasta.py (hunk 3 above)
Comment (wrightmhw) [inline screed/tests/test_fasta.py]: (do this check by hand and make sure the sequence is correct)

**Comment by wrightmhw** [comment]
Test this please.

**Comment by wrightmhw** [comment]
Test this please.

**Comment by wrightmhw** [comment]
Test this please.

**Comment by wrightmhw** [comment]
please test this

**Comment by wrightmhw** [comment]
test this please

**Comment by wrightmhw** [comment]
test this please.

**Comment by wrightmhw** [comment]
@mr-c I dont really understand Jenkins here, nor what is failing.
looks to me like `ERROR: Failure: ImportError (cannot import name Record)` is the problem, but I did not run into this error with any of my previous commits, and I can't seem to get Jenkins to retest this to see if it is still an issue or not.

**Comment by bocajnotnef** [comment]
retest this please

**Comment by bocajnotnef** [comment]
Hmm. Wonder if the Screed PR builder broke again.

**Comment by bocajnotnef** [comment]
Interesting. I attempted to force a manual build by passing in the SHA1 of @wrightmhw's last commit (` d52200a29be2e42a62846bae955dd8e5687e36ea`) and I'm getting an error that the builder can't find the revision to build....

**Comment by bocajnotnef** [comment]
Yep, builder broke.

**Comment by bocajnotnef** [comment]
I walked through the configs for screed and khmer and can't find any differences that would account for this, which makes this officially beyond my ken/above my paygrade. Sorry, @wrightmhw!

**Comment by wrightmhw** [comment]
@bocajnotnef I think it worked! Thank you for all of that!

File: screed/screedRecord.py [hunk 5]
```diff
@@ -32,6 +32,19 @@ def __len__(self):
     def keys(self):
         return self.d.keys()
 
+<<<<<<< HEAD
```
Comment (mr-c) [inline screed/screedRecord.py]: @bocajnotnef  ???

**Comment by bocajnotnef** [comment]
A couple of your tests are gonna start failing because one of the changes merged in were my chanegs to the screed test structure--some input files were moved to `screed/tests/test-data` and the khmer-like `get_temp_filename` was also implemented.

**Comment by luizirber** [comment]
@wrightmhw can I finish this?

**Comment by luizirber** [comment]
- [ ] Is it mergable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality, is it tested?
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  and `make doc` output. Use `make format` and manual fixing as needed.
- [ ] Is it documented in the ChangeLog?
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

**Comment by luizirber** [comment]
Ready for review and merge @bocajnotnef

**Comment by bocajnotnef** [comment]
merged #29

File: screed/screedRecord.py [hunk 6]
```diff
@@ -37,6 +34,19 @@ def __len__(self):
     def keys(self):
         return self.d.keys()
 
+    def __getitem__(self, idx):
+        if isinstance(idx, slice):
+            new_read = {k: self.d[k] for k in self.d
```
Comment (ctb) [inline screed/screedRecord.py]: Please simplify - this is silly. Why just do new_read = dict(self.d)??

Comment (ctb) [inline screed/screedRecord.py]: (I mean, whatever the equivalent is.)

File: screed/tests/test_fasta.py [hunk 7]
```diff
@@ -151,3 +151,18 @@ class FakeRecord(object):
...
+    shutil.copy(utils.get_test_data('test.fa'), testfa)
+
+    with screed.open(testfa) as index:
+        record = next(index)
+        trimmed = record[:10]
+
+    for k in set(record.keys()) - {'sequence'}:
+        assert trimmed[k] == record[k]
+
+    assert trimmed['sequence'] == record['sequence'][:10]
```
Comment (ctb) [inline screed/tests/test_fasta.py]: Let's add a few more slicing tests here, please.

**Comment by ctb** [comment]
Pls address comments, or I can submit a new PR, or whatever... just let me know which.

**Comment by luizirber** [comment]
@ctb I'm on it.

**Comment by ctb** [comment]
On Fri, Jun 05, 2015 at 11:11:28AM -0700, Luiz Irber wrote:
> @ctb I'm on it.

thanks!

**Pull request merged**